percent_encoding/src/lib.rs

```diff
@@ -79,7 +79,7 @@ const BITS_PER_CHUNK: usize = 8 * mem::size_of::<Chunk>();
 
 impl AsciiSet {
     /// An empty set.
-    pub const EMPTY: AsciiSet = AsciiSet {
+    pub const EMPTY: &'static AsciiSet = &AsciiSet {
```

Comment:
@joshka (continuing here to have a thread)

I don't think there's a reason it needs to be a reference, but given that the existing constants are references I think it makes sense to just have all of them be the same.

Changing the existing ones to not be references would be a breaking change, so this kinda seems like the only option to me.

Reply:
I don't really know the design decisions behind this well enough to give a useful answer. I'd defer to the library maintainers for more understanding / context on that. Making a constant that is a reference to a constant value instead of just the value seems just a little odd to me. It's also unlikely that this EMPTY const should ever be used except in some other constant expression, so it seems likely to me that a non-ref is still more correct here.

Rather than using AsRef as suggested, if I was fixing this up a bit to make it work with either values or refs, I'd add derived implementations for `Clone` and `Copy`, `impl Into<AsciiSet> for &'_ AsciiSet { fn into(self) -> AsciiSet { *self } }`, and then change the methods to accept `Into<AsciiSet>` and the PercentEncoding struct to just store the value instead of a ref. This would be both backward compatible and obvious. The caveat to this is I'm unsure if this code is called in some performance sensitive situation however (e.g where the nanoseconds matter), It's 16 bytes of memory copied so I'd hope a copy would be fine at some general level. It may not be if this is used in a super high volume scenario (e.g. a firewall or proxy). There aren't any benchmarks to imply that this would have some high perf needs.

Reply:
The performance shouldn't matter too much either way, the more pressing thing for static-vs-const is the instruction size bloat if the const gets duplicated everywhere, but this is a small const. The compiler is also able to optimize in both ways at times. I'd say that the reference is slightly better just because of consistency, but for smaller consts in general a straight up const is better overall.

Reply:
https://github.com/servo/rust-url/pull/976 does the refactoring mentioned